// agentwallet-sdk/dist/identity/erc8004.js
/**
 * ERC-8004: Trustless Agents — Identity Registry Integration
 *
 * Implements the Identity Registry component of ERC-8004, which provides every
 * AI agent with a portable, censorship-resistant on-chain identifier using ERC-721
 * with URIStorage extension.
 *
 * Spec: https://eips.ethereum.org/EIPS/eip-8004
 * Status: DRAFT (August 2025). No official mainnet singleton deployed yet.
 * Key Principle: Non-custodial. All signing happens locally via WalletClient.
 */
import { createPublicClient, getContract, http, } from 'viem';
import { base, baseSepolia, mainnet, arbitrum, polygon } from 'viem/chains';
// ─── ABI ─────────────────────────────────────────────────────────────────────
export const ERC8004IdentityRegistryAbi = [
    {
        name: 'register', type: 'function', stateMutability: 'nonpayable',
        inputs: [{ name: 'agentURI', type: 'string' }],
        outputs: [{ name: 'agentId', type: 'uint256' }],
    },
    {
        name: 'registerWithMetadata', type: 'function', stateMutability: 'nonpayable',
        inputs: [
            { name: 'agentURI', type: 'string' },
            { name: 'metadata', type: 'tuple[]', components: [
                    { name: 'metadataKey', type: 'string' },
                    { name: 'metadataValue', type: 'bytes' },
                ] },
        ],
        outputs: [{ name: 'agentId', type: 'uint256' }],
    },
    {
        name: 'registerEmpty', type: 'function', stateMutability: 'nonpayable',
        inputs: [], outputs: [{ name: 'agentId', type: 'uint256' }],
    },
    {
        name: 'setAgentURI', type: 'function', stateMutability: 'nonpayable',
        inputs: [{ name: 'agentId', type: 'uint256' }, { name: 'newURI', type: 'string' }],
        outputs: [],
    },
    {
        name: 'tokenURI', type: 'function', stateMutability: 'view',
        inputs: [{ name: 'agentId', type: 'uint256' }],
        outputs: [{ name: '', type: 'string' }],
    },
    {
        name: 'getMetadata', type: 'function', stateMutability: 'view',
        inputs: [{ name: 'agentId', type: 'uint256' }, { name: 'metadataKey', type: 'string' }],
        outputs: [{ name: '', type: 'bytes' }],
    },
    {
        name: 'setMetadata', type: 'function', stateMutability: 'nonpayable',
        inputs: [
            { name: 'agentId', type: 'uint256' },
            { name: 'metadataKey', type: 'string' },
            { name: 'metadataValue', type: 'bytes' },
        ],
        outputs: [],
    },
    {
        name: 'getAgentWallet', type: 'function', stateMutability: 'view',
        inputs: [{ name: 'agentId', type: 'uint256' }],
        outputs: [{ name: '', type: 'address' }],
    },
    {
        name: 'setAgentWallet', type: 'function', stateMutability: 'nonpayable',
        inputs: [
            { name: 'agentId', type: 'uint256' }, { name: 'newWallet', type: 'address' },
            { name: 'deadline', type: 'uint256' }, { name: 'signature', type: 'bytes' },
        ],
        outputs: [],
    },
    {
        name: 'unsetAgentWallet', type: 'function', stateMutability: 'nonpayable',
        inputs: [{ name: 'agentId', type: 'uint256' }], outputs: [],
    },
    {
        name: 'ownerOf', type: 'function', stateMutability: 'view',
        inputs: [{ name: 'tokenId', type: 'uint256' }],
        outputs: [{ name: '', type: 'address' }],
    },
    {
        name: 'Registered', type: 'event', inputs: [
            { name: 'agentId', type: 'uint256', indexed: true },
            { name: 'agentURI', type: 'string', indexed: false },
            { name: 'owner', type: 'address', indexed: true },
        ],
    },
    {
        name: 'URIUpdated', type: 'event', inputs: [
            { name: 'agentId', type: 'uint256', indexed: true },
            { name: 'newURI', type: 'string', indexed: false },
            { name: 'updatedBy', type: 'address', indexed: true },
        ],
    },
    {
        name: 'MetadataSet', type: 'event', inputs: [
            { name: 'agentId', type: 'uint256', indexed: true },
            { name: 'indexedMetadataKey', type: 'string', indexed: true },
            { name: 'metadataKey', type: 'string', indexed: false },
            { name: 'metadataValue', type: 'bytes', indexed: false },
        ],
    },
];
// ─── Constants ───────────────────────────────────────────────────────────────
export const METADATA_KEYS = {
    AGENT_WALLET: 'agentWallet',
    MODEL: 'agentModel',
    MODEL_PROVIDER: 'agentModelProvider',
    VERSION: 'agentVersion',
    CAPABILITIES: 'agentCapabilities',
    FRAMEWORK: 'agentFramework',
};
export const REGISTRATION_FILE_TYPE = 'https://eips.ethereum.org/EIPS/eip-8004#registration-v1';
export const KNOWN_REGISTRY_ADDRESSES = {
    'base-sepolia': '0x0000000000000000000000000000000000000000',
};
const CHAINS = {
    base, 'base-sepolia': baseSepolia, ethereum: mainnet, arbitrum, polygon,
};
const REGISTERED_TOPIC = '0x6f3c9b7e8c3a5a5f2c3f9b7e8c3a5a5f2c3f9b7e8c3a5a5f2c3f9b7e8c3a5a';
// ─── Client ──────────────────────────────────────────────────────────────────
export class ERC8004Client {
    constructor(config) {
        this.config = config;
        const chain = CHAINS[config.chain];
        if (!chain)
            throw new Error(`ERC8004Client: Unsupported chain "${config.chain}"`);
        this.chain = chain;
        this.publicClient = createPublicClient({ chain, transport: http(config.rpcUrl) });
    }
    getReadContract() {
        return getContract({
            address: this.config.registryAddress,
            abi: ERC8004IdentityRegistryAbi,
            client: this.publicClient,
        });
    }
    getWriteContract(walletClient) {
        return getContract({
            address: this.config.registryAddress,
            abi: ERC8004IdentityRegistryAbi,
            client: { public: this.publicClient, wallet: walletClient },
        });
    }
    async registerAgent(walletClient, metadata, agentURI, onChainMetadata) {
        if (!walletClient.account)
            throw new Error('ERC8004Client: WalletClient has no account');
        const registrationFile = { type: REGISTRATION_FILE_TYPE, ...metadata };
        const resolvedURI = agentURI ?? buildDataURI(registrationFile);
        const contract = this.getWriteContract(walletClient);
        let txHash;
        if (onChainMetadata && Object.keys(onChainMetadata).length > 0) {
            const entries = encodeMetadataEntries(onChainMetadata);
            txHash = await contract.write.registerWithMetadata([resolvedURI, entries], {
                account: walletClient.account, chain: this.chain,
            });
        }
        else {
            txHash = await contract.write.register([resolvedURI], {
                account: walletClient.account, chain: this.chain,
            });
        }
        let agentId = null;
        try {
            const receipt = await this.publicClient.waitForTransactionReceipt({ hash: txHash });
            const registeredEvent = receipt.logs.find((log) => log.address.toLowerCase() === this.config.registryAddress.toLowerCase() &&
                log.topics[0] === REGISTERED_TOPIC);
            if (registeredEvent?.topics[1]) {
                agentId = BigInt(registeredEvent.topics[1]);
            }
        }
        catch {
            // agentId remains null
        }
        return { txHash, agentId };
    }
    async lookupAgentIdentity(agentId) {
        const contract = this.getReadContract();
        const [owner, agentURI, agentWallet] = await Promise.all([
            contract.read.ownerOf([agentId]),
            contract.read.tokenURI([agentId]),
            contract.read.getAgentWallet([agentId]),
        ]);
        let registrationFile = null;
        try {
            registrationFile = await resolveAgentURI(agentURI);
        }
        catch { /* unavailable */ }
        const modelMetadata = await this.readModelMetadata(agentId);
        return {
            agentId, owner: owner, agentURI: agentURI,
            agentWallet: agentWallet, registrationFile, modelMetadata,
        };
    }
    async lookupAgentByOwner(ownerAddress, fromBlock = 0n) {
        const logs = await this.publicClient.getLogs({
            address: this.config.registryAddress,
            event: {
                type: 'event', name: 'Registered',
                inputs: [
                    { name: 'agentId', type: 'uint256', indexed: true },
                    { name: 'agentURI', type: 'string', indexed: false },
                    { name: 'owner', type: 'address', indexed: true },
                ],
            },
            args: { owner: ownerAddress },
            fromBlock, toBlock: 'latest',
        });
        if (logs.length === 0)
            return null;
        const latestLog = logs[logs.length - 1];
        const agentId = BigInt(latestLog.args?.agentId ?? 0);
        return this.lookupAgentIdentity(agentId);
    }
    async updateAgentURI(walletClient, agentId, newURI) {
        if (!walletClient.account)
            throw new Error('ERC8004Client: WalletClient has no account');
        const contract = this.getWriteContract(walletClient);
        return contract.write.setAgentURI([agentId, newURI], {
            account: walletClient.account, chain: this.chain,
        });
    }
    async getOnChainMetadata(agentId, key) {
        const contract = this.getReadContract();
        const value = await contract.read.getMetadata([agentId, key]);
        return value === '0x' || value === null ? null : value;
    }
    async setOnChainMetadata(walletClient, agentId, key, value) {
        if (!walletClient.account)
            throw new Error('ERC8004Client: WalletClient has no account');
        if (key === METADATA_KEYS.AGENT_WALLET) {
            throw new Error('ERC8004Client: "agentWallet" is reserved — use setAgentWallet() instead');
        }
        const contract = this.getWriteContract(walletClient);
        const hexValue = stringToHex(value);
        return contract.write.setMetadata([agentId, key, hexValue], {
            account: walletClient.account, chain: this.chain,
        });
    }
    async setModelMetadata(walletClient, agentId, model) {
        const hashes = [];
        const entries = [];
        if (model.model)
            entries.push([METADATA_KEYS.MODEL, model.model]);
        if (model.provider)
            entries.push([METADATA_KEYS.MODEL_PROVIDER, model.provider]);
        if (model.version)
            entries.push([METADATA_KEYS.VERSION, model.version]);
        if (model.framework)
            entries.push([METADATA_KEYS.FRAMEWORK, model.framework]);
        if (model.capabilities?.length)
            entries.push([METADATA_KEYS.CAPABILITIES, JSON.stringify(model.capabilities)]);
        for (const [key, val] of entries) {
            const hash = await this.setOnChainMetadata(walletClient, agentId, key, val);
            hashes.push(hash);
        }
        return hashes;
    }
    async readModelMetadata(agentId) {
        const keys = [METADATA_KEYS.MODEL, METADATA_KEYS.MODEL_PROVIDER, METADATA_KEYS.VERSION, METADATA_KEYS.FRAMEWORK, METADATA_KEYS.CAPABILITIES];
        const values = await Promise.all(keys.map((k) => this.getOnChainMetadata(agentId, k)));
        const [modelHex, providerHex, versionHex, frameworkHex, capabilitiesHex] = values;
        if (!modelHex && !providerHex && !versionHex && !frameworkHex && !capabilitiesHex)
            return null;
        const result = {};
        if (modelHex)
            result.model = hexToString(modelHex);
        if (providerHex)
            result.provider = hexToString(providerHex);
        if (versionHex)
            result.version = hexToString(versionHex);
        if (frameworkHex)
            result.framework = hexToString(frameworkHex);
        if (capabilitiesHex) {
            try {
                result.capabilities = JSON.parse(hexToString(capabilitiesHex));
            }
            catch {
                result.capabilities = [];
            }
        }
        return result;
    }
    async getAgentWallet(agentId) {
        const contract = this.getReadContract();
        return contract.read.getAgentWallet([agentId]);
    }
    async setAgentWallet(walletClient, agentId, newWallet, deadline, signature) {
        if (!walletClient.account)
            throw new Error('ERC8004Client: WalletClient has no account');
        const contract = this.getWriteContract(walletClient);
        return contract.write.setAgentWallet([agentId, newWallet, deadline, signature], {
            account: walletClient.account, chain: this.chain,
        });
    }
    async unsetAgentWallet(walletClient, agentId) {
        if (!walletClient.account)
            throw new Error('ERC8004Client: WalletClient has no account');
        const contract = this.getWriteContract(walletClient);
        return contract.write.unsetAgentWallet([agentId], {
            account: walletClient.account, chain: this.chain,
        });
    }
}
// ─── Standalone helpers ───────────────────────────────────────────────────────
export function buildDataURI(registrationFile) {
    const payload = { ...registrationFile, type: REGISTRATION_FILE_TYPE };
    const json = JSON.stringify(payload);
    const b64 = btoa(unescape(encodeURIComponent(json)));
    return `data:application/json;base64,${b64}`;
}
export function parseDataURI(uri) {
    if (uri.startsWith('data:application/json;base64,')) {
        const b64 = uri.replace('data:application/json;base64,', '');
        const json = decodeURIComponent(escape(atob(b64)));
        return JSON.parse(json);
    }
    if (uri.startsWith('{'))
        return JSON.parse(uri);
    throw new Error(`parseDataURI: Cannot parse URI scheme: ${uri.substring(0, 50)}`);
}
export async function resolveAgentURI(uri) {
    if (uri.startsWith('data:'))
        return parseDataURI(uri);
    if (uri.startsWith('https://') || uri.startsWith('http://')) {
        const response = await fetch(uri);
        if (!response.ok)
            throw new Error(`resolveAgentURI: HTTP ${response.status} for ${uri}`);
        return response.json();
    }
    throw new Error(`resolveAgentURI: Unsupported URI scheme: ${uri}`);
}
export function validateRegistrationFile(file) {
    const errors = [];
    if (file.type !== REGISTRATION_FILE_TYPE)
        errors.push(`type must be "${REGISTRATION_FILE_TYPE}"`);
    if (!file.name || typeof file.name !== 'string' || file.name.trim() === '')
        errors.push('name is required');
    if (!file.description || typeof file.description !== 'string')
        errors.push('description is required');
    if (file.services) {
        file.services.forEach((svc, i) => {
            if (!svc.name)
                errors.push(`services[${i}].name is required`);
            if (!svc.endpoint)
                errors.push(`services[${i}].endpoint is required`);
        });
    }
    if (file.registrations) {
        file.registrations.forEach((reg, i) => {
            if (reg.agentId === undefined)
                errors.push(`registrations[${i}].agentId is required`);
            if (!reg.agentRegistry)
                errors.push(`registrations[${i}].agentRegistry is required`);
        });
    }
    return errors;
}
export function formatAgentRegistry(chainId, registryAddress) {
    return `eip155:${chainId}:${registryAddress}`;
}
// ─── Internal utilities ───────────────────────────────────────────────────────
function stringToHex(str) {
    const bytes = new TextEncoder().encode(str);
    const hexStr = Array.from(bytes).map((b) => b.toString(16).padStart(2, '0')).join('');
    return ('0x' + hexStr);
}
function hexToString(hex) {
    const stripped = hex.startsWith('0x') ? hex.slice(2) : hex;
    const bytes = new Uint8Array(stripped.match(/.{1,2}/g).map((byte) => parseInt(byte, 16)));
    return new TextDecoder('utf-8').decode(bytes);
}
function encodeMetadataEntries(entries) {
    return Object.entries(entries).map(([key, value]) => ({
        metadataKey: key,
        metadataValue: stringToHex(value),
    }));
}
