// agentwallet-sdk/dist/bridge/client.js
// BridgeModule — CCTP V2 cross-chain USDC bridge for AgentWallet
// Non-custodial: all signing is done locally via the agent's WalletClient.
import { createPublicClient, http, keccak256, pad, getContract, } from 'viem';
import { base, mainnet, optimism, arbitrum } from 'viem/chains';
import { TokenMessengerV2Abi, MessageTransmitterV2Abi, ERC20BridgeAbi, } from './abis.js';
import { CCTP_DOMAIN_IDS, BRIDGE_CHAIN_IDS, USDC_CONTRACT, TOKEN_MESSENGER_V2, MESSAGE_TRANSMITTER_V2, FINALITY_THRESHOLD, CIRCLE_ATTESTATION_API, MAX_ATTESTATION_POLLS, ATTESTATION_POLL_INTERVAL_MS, } from './types.js';
const VIEM_CHAINS = { base, ethereum: mainnet, optimism, arbitrum };
export class BridgeModule {
    constructor(walletClient, fromChain = 'base', options = {}) {
        if (!walletClient.account) {
            throw new BridgeError('NO_WALLET_CLIENT', 'WalletClient must have an account attached. Use privateKeyToAccount() or similar to attach a signer.');
        }
        this.walletClient = walletClient;
        this.fromChain = fromChain;
        this.fromRpcUrl = options.rpcUrl;
        this.publicClient = createPublicClient({
            chain: VIEM_CHAINS[fromChain],
            transport: http(options.rpcUrl),
        });
    }
    async bridge(amount, toChain, options = {}) {
        const startMs = Date.now();
        this.validateBridgeParams(amount, toChain);
        const account = this.walletClient.account;
        const recipient = options.destinationAddress ?? account.address;
        const minFinalityThreshold = options.minFinalityThreshold ?? FINALITY_THRESHOLD.FAST;
        const maxFee = options.maxFee ?? 0n;
        const attestationApiUrl = options.attestationApiUrl ?? CIRCLE_ATTESTATION_API;
        await this.approveUsdc(amount);
        const burnResult = await this.depositForBurn(amount, toChain, recipient, minFinalityThreshold, maxFee);
        const attestation = await this.pollForAttestation(burnResult.messageHash, this.fromChain, attestationApiUrl);
        const mintTxHash = await this.receiveMessage(burnResult.messageBytes, attestation, toChain, options.destinationRpcUrl);
        return {
            burnTxHash: burnResult.burnTxHash,
            mintTxHash,
            amount,
            fromChain: this.fromChain,
            toChain,
            recipient,
            nonce: burnResult.nonce,
            elapsedMs: Date.now() - startMs,
        };
    }
    async burn(amount, toChain, options = {}) {
        this.validateBridgeParams(amount, toChain);
        const account = this.walletClient.account;
        const recipient = options.destinationAddress ?? account.address;
        const minFinalityThreshold = options.minFinalityThreshold ?? FINALITY_THRESHOLD.FAST;
        const maxFee = options.maxFee ?? 0n;
        await this.approveUsdc(amount);
        return this.depositForBurn(amount, toChain, recipient, minFinalityThreshold, maxFee);
    }
    async waitForAttestation(messageHash, apiUrl = CIRCLE_ATTESTATION_API) {
        return this.pollForAttestation(messageHash, this.fromChain, apiUrl);
    }
    async mint(messageBytes, attestation, toChain, destinationRpcUrl) {
        return this.receiveMessage(messageBytes, attestation, toChain, destinationRpcUrl);
    }
    async getUsdcBalance() {
        const account = this.walletClient.account;
        const usdc = getContract({
            address: USDC_CONTRACT[this.fromChain],
            abi: ERC20BridgeAbi,
            client: this.publicClient,
        });
        return usdc.read.balanceOf([account.address]);
    }
    async getUsdcAllowance() {
        const account = this.walletClient.account;
        const usdc = getContract({
            address: USDC_CONTRACT[this.fromChain],
            abi: ERC20BridgeAbi,
            client: this.publicClient,
        });
        return usdc.read.allowance([account.address, TOKEN_MESSENGER_V2[this.fromChain]]);
    }
    async approveUsdc(amount) {
        const account = this.walletClient.account;
        const spender = TOKEN_MESSENGER_V2[this.fromChain];
        const usdcAddress = USDC_CONTRACT[this.fromChain];
        const usdc = getContract({ address: usdcAddress, abi: ERC20BridgeAbi, client: this.publicClient });
        const currentAllowance = await usdc.read.allowance([account.address, spender]);
        if (currentAllowance >= amount)
            return;
        const usdcWrite = getContract({
            address: usdcAddress,
            abi: ERC20BridgeAbi,
            client: { public: this.publicClient, wallet: this.walletClient },
        });
        const approveTxHash = await usdcWrite.write.approve([spender, amount], {
            account,
            chain: VIEM_CHAINS[this.fromChain],
        });
        const approveReceipt = await this.publicClient.waitForTransactionReceipt({ hash: approveTxHash });
        if (approveReceipt.status !== 'success') {
            throw new BridgeError('INSUFFICIENT_ALLOWANCE', `USDC approve failed (tx: ${approveTxHash}).`);
        }
    }
    async depositForBurn(amount, toChain, recipient, minFinalityThreshold, maxFee) {
        const account = this.walletClient.account;
        const destinationDomain = CCTP_DOMAIN_IDS[toChain];
        const messengerAddress = TOKEN_MESSENGER_V2[this.fromChain];
        const usdcAddress = USDC_CONTRACT[this.fromChain];
        const mintRecipient = pad(recipient, { size: 32 });
        const destinationCaller = pad('0x0', { size: 32 });
        const messenger = getContract({
            address: messengerAddress,
            abi: TokenMessengerV2Abi,
            client: { public: this.publicClient, wallet: this.walletClient },
        });
        let burnTxHash;
        try {
            burnTxHash = await messenger.write.depositForBurn([amount, destinationDomain, mintRecipient, usdcAddress, destinationCaller, maxFee, minFinalityThreshold], { account, chain: VIEM_CHAINS[this.fromChain] });
        }
        catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            throw new BridgeError('BURN_FAILED', `CCTP depositForBurn failed: ${msg}.`);
        }
        const receipt = await this.publicClient.waitForTransactionReceipt({ hash: burnTxHash });
        if (receipt.status !== 'success') {
            throw new BridgeError('BURN_FAILED', `depositForBurn transaction reverted (tx: ${burnTxHash}).`);
        }
        const { messageBytes, messageHash, nonce } = this.extractMessageSent(receipt);
        return { burnTxHash, nonce, messageHash, messageBytes, sourceDomain: CCTP_DOMAIN_IDS[this.fromChain], destinationDomain };
    }
    extractMessageSent(receipt) {
        const MESSAGE_SENT_TOPIC = '0x8c5261668696ce22758910d05bab8f186d6eb247ceac2af2e82c7dc17669b036';
        for (const log of receipt.logs) {
            if (log.topics[0]?.toLowerCase() === MESSAGE_SENT_TOPIC.toLowerCase()) {
                const rawData = log.data;
                if (rawData.length < 130)
                    continue;
                const dataHex = rawData.slice(2);
                const lengthHex = dataHex.slice(64, 128);
                const messageLength = parseInt(lengthHex, 16);
                if (messageLength === 0)
                    continue;
                const messageBytesHex = dataHex.slice(128, 128 + messageLength * 2);
                const messageBytes = ('0x' + messageBytesHex);
                const messageHash = keccak256(messageBytes);
                const nonceBytesHex = messageBytesHex.slice(24, 40);
                const nonce = BigInt('0x' + nonceBytesHex);
                return { messageBytes, messageHash, nonce };
            }
        }
        throw new BridgeError('BURN_FAILED', 'Could not find MessageSent event in burn transaction receipt.');
    }
    async pollForAttestation(messageHash, fromChain, apiUrl) {
        const sourceDomain = CCTP_DOMAIN_IDS[fromChain];
        const url = `${apiUrl}/v2/messages/${sourceDomain}/${messageHash}`;
        for (let attempt = 0; attempt < MAX_ATTESTATION_POLLS; attempt++) {
            let response;
            try {
                const res = await fetch(url, { headers: { Accept: 'application/json' } });
                if (!res.ok) {
                    if (res.status === 404) {
                        await this.sleep(ATTESTATION_POLL_INTERVAL_MS);
                        continue;
                    }
                    const body = await res.text().catch(() => '');
                    throw new BridgeError('ATTESTATION_ERROR', `Circle API returned HTTP ${res.status}: ${body}.`);
                }
                response = await res.json();
            }
            catch (err) {
                if (err instanceof BridgeError)
                    throw err;
                const msg = err instanceof Error ? err.message : String(err);
                throw new BridgeError('ATTESTATION_ERROR', `Failed to reach Circle IRIS API: ${msg}.`);
            }
            if (response.status === 'complete' && response.attestation)
                return response.attestation;
            if (response.status === 'error') {
                throw new BridgeError('ATTESTATION_ERROR', `Circle attestation failed: ${response.error ?? 'unknown error'}.`);
            }
            await this.sleep(ATTESTATION_POLL_INTERVAL_MS);
        }
        throw new BridgeError('ATTESTATION_TIMEOUT', `Attestation not received after ${MAX_ATTESTATION_POLLS} attempts. Message hash: ${messageHash}.`);
    }
    async receiveMessage(messageBytes, attestation, toChain, destinationRpcUrl) {
        const account = this.walletClient.account;
        const transmitterAddress = MESSAGE_TRANSMITTER_V2[toChain];
        const destChain = VIEM_CHAINS[toChain];
        const destPublicClient = createPublicClient({ chain: destChain, transport: http(destinationRpcUrl) });
        const transmitter = getContract({
            address: transmitterAddress,
            abi: MessageTransmitterV2Abi,
            client: { public: destPublicClient, wallet: this.walletClient },
        });
        let mintTxHash;
        try {
            mintTxHash = await transmitter.write.receiveMessage([messageBytes, attestation], { account, chain: destChain });
        }
        catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            throw new BridgeError('MINT_FAILED', `CCTP receiveMessage failed on ${toChain}: ${msg}.`);
        }
        const mintReceipt = await destPublicClient.waitForTransactionReceipt({ hash: mintTxHash });
        if (mintReceipt.status !== 'success') {
            throw new BridgeError('MINT_FAILED', `receiveMessage reverted on ${toChain} (tx: ${mintTxHash}).`);
        }
        return mintTxHash;
    }
    validateBridgeParams(amount, toChain) {
        if (amount <= 0n) {
            throw new BridgeError('INVALID_AMOUNT', `Bridge amount must be greater than 0. Received: ${amount}.`);
        }
        if (!(toChain in CCTP_DOMAIN_IDS)) {
            throw new BridgeError('UNSUPPORTED_CHAIN', `Chain '${toChain}' is not supported.`);
        }
        if (toChain === this.fromChain) {
            throw new BridgeError('UNSUPPORTED_CHAIN', `Source and destination chains must be different.`);
        }
    }
    sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
}
export class BridgeError extends Error {
    constructor(code, message) {
        super(`[BridgeModule:${code}] ${message}`);
        this.code = code;
        this.name = 'BridgeError';
    }
}
export function createBridge(walletClient, fromChain = 'base', options = {}) {
    return new BridgeModule(walletClient, fromChain, options);
}
export { CCTP_DOMAIN_IDS, BRIDGE_CHAIN_IDS, USDC_CONTRACT, TOKEN_MESSENGER_V2, MESSAGE_TRANSMITTER_V2, FINALITY_THRESHOLD, };
