// agentwallet-sdk/dist/swap/SwapModule.js
// SwapModule — Uniswap V3 token swap aggregator with 0.875% protocol fee
import { encodeFunctionData, } from 'viem';
import { UniswapV3RouterAbi, UniswapV3QuoterV2Abi, ERC20Abi } from './abi.js';
import { PROTOCOL_FEE_BPS, DEFAULT_SLIPPAGE_BPS, UNISWAP_V3_BASE, } from './types.js';
const FEE_TIERS = [100, 500, 3000, 10000];
/** Calculate protocol fee: floor(amount * feeBps / 100_000) */
export function calcProtocolFee(amount, feeBps) {
    return (amount * BigInt(feeBps)) / 100000n;
}
/** Apply slippage: floor(amount * (10_000 - slippageBps) / 10_000) */
export function applySlippage(amount, slippageBps) {
    return (amount * BigInt(10000 - slippageBps)) / 10000n;
}
/** Encode deadline: block.timestamp + deadlineSecs */
export function calcDeadline(deadlineSecs) {
    return BigInt(Math.floor(Date.now() / 1000) + deadlineSecs);
}
export class SwapModule {
    constructor(publicClient, walletClient, accountAddress, config) {
        this.publicClient = publicClient;
        this.walletClient = walletClient;
        this.accountAddress = accountAddress;
        this.config = {
            routerAddress: UNISWAP_V3_BASE.ROUTER,
            quoterAddress: UNISWAP_V3_BASE.QUOTER_V2,
            feeBps: PROTOCOL_FEE_BPS,
            feeWallet: accountAddress,
            ...config,
        };
    }
    async getQuote(tokenIn, tokenOut, amountIn, options = {}) {
        const slippageBps = options.slippageBps ?? DEFAULT_SLIPPAGE_BPS;
        const tiersToTry = options.feeTiers ?? FEE_TIERS;
        const feeAmount = calcProtocolFee(amountIn, this.config.feeBps);
        const amountInNet = amountIn - feeAmount;
        if (amountInNet <= 0n) {
            throw new Error(`SwapModule: amountIn (${amountIn}) is too small — fee (${feeAmount}) exceeds input`);
        }
        let bestQuote = null;
        for (const fee of tiersToTry) {
            try {
                const result = await this.publicClient.readContract({
                    address: this.config.quoterAddress,
                    abi: UniswapV3QuoterV2Abi,
                    functionName: 'quoteExactInputSingle',
                    args: [{ tokenIn, tokenOut, amount: amountInNet, fee, sqrtPriceLimitX96: 0n }],
                });
                const [amountOut, , , gasEstimate] = result;
                if (!bestQuote || amountOut > bestQuote.amountOut) {
                    bestQuote = { amountOut, gasEstimate, feeTier: fee };
                }
            }
            catch {
                // Pool doesn't exist for this fee tier
            }
        }
        if (!bestQuote) {
            throw new Error(`SwapModule: No Uniswap V3 pool found for ${tokenIn} → ${tokenOut} on Base.`);
        }
        const amountOutMinimum = applySlippage(bestQuote.amountOut, slippageBps);
        const effectiveRate = amountInNet > 0n ? Number(bestQuote.amountOut) / Number(amountInNet) : 0;
        return {
            tokenIn, tokenOut, amountInRaw: amountIn, amountInNet, feeAmount,
            amountOut: bestQuote.amountOut, amountOutMinimum,
            poolFeeTier: bestQuote.feeTier, effectiveRate, gasEstimate: bestQuote.gasEstimate,
        };
    }
    async ensureApproval(token, spender, amount) {
        const currentAllowance = (await this.publicClient.readContract({
            address: token, abi: ERC20Abi, functionName: 'allowance',
            args: [this.accountAddress, spender],
        }));
        if (currentAllowance >= amount)
            return undefined;
        const MAX_UINT256 = 115792089237316195423570985008687907853269984665640564039457584007913129639935n;
        const account = this.walletClient.account;
        if (!account)
            throw new Error('SwapModule: walletClient has no account');
        const approveData = encodeFunctionData({
            abi: ERC20Abi, functionName: 'approve', args: [spender, MAX_UINT256],
        });
        return this.walletClient.sendTransaction({
            account, to: token, data: approveData, chain: this.walletClient.chain ?? null,
        });
    }
    async transferFee(token, feeAmount, feeWallet) {
        const account = this.walletClient.account;
        if (!account)
            throw new Error('SwapModule: walletClient has no account');
        const transferData = encodeFunctionData({
            abi: ERC20Abi, functionName: 'transfer', args: [feeWallet, feeAmount],
        });
        return this.walletClient.sendTransaction({
            account, to: token, data: transferData, chain: this.walletClient.chain ?? null,
        });
    }
    async swap(tokenIn, tokenOut, amountIn, options = {}) {
        const slippageBps = options.slippageBps ?? DEFAULT_SLIPPAGE_BPS;
        const deadlineSecs = options.deadlineSecs ?? 300;
        const feeWallet = options.feeWallet ?? this.config.feeWallet;
        const quote = await this.getQuote(tokenIn, tokenOut, amountIn, { slippageBps, feeTiers: options.feeTiers });
        if (quote.gasEstimate > 2000000n) {
            console.warn(`SwapModule: High gas estimate (${quote.gasEstimate}). Pool may be illiquid.`);
        }
        const approvalTxHash = await this.ensureApproval(tokenIn, this.config.routerAddress, amountIn);
        const approvalRequired = approvalTxHash !== undefined;
        let feeTxHash;
        if (quote.feeAmount > 0n && feeWallet !== this.accountAddress) {
            feeTxHash = await this.transferFee(tokenIn, quote.feeAmount, feeWallet);
        }
        const account = this.walletClient.account;
        if (!account)
            throw new Error('SwapModule: walletClient has no account');
        const deadline = calcDeadline(deadlineSecs);
        const swapData = encodeFunctionData({
            abi: UniswapV3RouterAbi,
            functionName: 'exactInputSingle',
            args: [{
                    tokenIn, tokenOut, fee: quote.poolFeeTier, recipient: this.accountAddress,
                    amountIn: quote.amountInNet, amountOutMinimum: quote.amountOutMinimum, sqrtPriceLimitX96: 0n,
                }],
        });
        const txHash = await this.walletClient.sendTransaction({
            account, to: this.config.routerAddress, data: swapData, chain: this.walletClient.chain ?? null,
        });
        return { txHash, feeTxHash, quote, approvalRequired, approvalTxHash };
    }
    getConfig() {
        return { ...this.config };
    }
    setFeeWallet(address) {
        this.config.feeWallet = address;
    }
}
export function attachSwap(wallet, config) {
    const swapModule = new SwapModule(wallet.publicClient, wallet.walletClient, wallet.address, config);
    return {
        ...wallet,
        swapModule,
        swap: swapModule.swap.bind(swapModule),
        getQuote: swapModule.getQuote.bind(swapModule),
    };
}
