// agentwallet-sdk/dist/x402/budget.js
/**
 * [MAX-ADDED] In-memory budget tracker for x402 payments.
 * Enforces per-service caps, daily limits, and per-request maximums.
 * On-chain spend limits are ALSO enforced by the AgentWallet contract —
 * this is an additional client-side layer for granular service-level control.
 */
export class X402BudgetTracker {
    constructor(config = {}) {
        this.serviceBudgets = new Map();
        this.dailySpend = new Map(); // service -> today's total
        this.globalDailySpend = 0n;
        this.transactionLog = [];
        this.globalDailyLimit = config.globalDailyLimit ?? BigInt('0xFFFFFFFFFFFFFFFFFFFFFFFF'); // effectively unlimited
        this.globalPerRequestMax = config.globalPerRequestMax ?? BigInt('0xFFFFFFFFFFFFFFFFFFFFFFFF');
        this.dailyResetTimestamp = this.startOfDay();
        if (config.serviceBudgets) {
            for (const budget of config.serviceBudgets) {
                this.serviceBudgets.set(budget.service, budget);
            }
        }
    }
    /**
     * Check if a payment is within budget limits.
     * Returns { allowed: true } or { allowed: false, reason: string }.
     */
    checkBudget(service, amount) {
        this.maybeResetDaily();
        // Global per-request check
        if (amount > this.globalPerRequestMax) {
            return { allowed: false, reason: `Amount ${amount} exceeds global per-request max ${this.globalPerRequestMax}` };
        }
        // Global daily check
        if (this.globalDailySpend + amount > this.globalDailyLimit) {
            return { allowed: false, reason: `Would exceed global daily limit ${this.globalDailyLimit}` };
        }
        // Service-specific checks
        const budget = this.findServiceBudget(service);
        if (budget) {
            if (amount > budget.maxPerRequest) {
                return { allowed: false, reason: `Amount ${amount} exceeds service per-request max ${budget.maxPerRequest} for ${service}` };
            }
            const serviceDailySpend = this.dailySpend.get(service) ?? 0n;
            if (serviceDailySpend + amount > budget.dailyLimit) {
                return { allowed: false, reason: `Would exceed daily limit ${budget.dailyLimit} for ${service}` };
            }
        }
        return { allowed: true };
    }
    /**
     * Record a completed payment.
     */
    recordPayment(log) {
        this.maybeResetDaily();
        this.transactionLog.push(log);
        if (log.success) {
            const service = log.service;
            this.dailySpend.set(service, (this.dailySpend.get(service) ?? 0n) + log.amount);
            this.globalDailySpend += log.amount;
        }
    }
    /**
     * Get transaction history, optionally filtered.
     */
    getTransactionLog(filter) {
        let logs = this.transactionLog;
        if (filter?.service) {
            logs = logs.filter(l => l.service === filter.service);
        }
        if (filter?.since) {
            logs = logs.filter(l => l.timestamp >= filter.since);
        }
        return logs;
    }
    /**
     * Get current daily spend summary.
     */
    getDailySpendSummary() {
        this.maybeResetDaily();
        const byService = {};
        for (const [service, amount] of this.dailySpend) {
            byService[service] = amount;
        }
        return {
            global: this.globalDailySpend,
            byService,
            resetsAt: this.dailyResetTimestamp + 86400,
        };
    }
    /**
     * Add or update a service budget at runtime.
     */
    setServiceBudget(budget) {
        this.serviceBudgets.set(budget.service, budget);
    }
    // ─── Internals ───
    findServiceBudget(service) {
        // Exact match first, then wildcard
        return this.serviceBudgets.get(service) ?? this.serviceBudgets.get('*');
    }
    maybeResetDaily() {
        const now = this.startOfDay();
        if (now > this.dailyResetTimestamp) {
            this.dailySpend.clear();
            this.globalDailySpend = 0n;
            this.dailyResetTimestamp = now;
        }
    }
    startOfDay() {
        const now = Math.floor(Date.now() / 1000);
        return now - (now % 86400);
    }
}
